core/validatorapi/validatorapi.go

```diff
@@ -960,27 +960,66 @@ func (c Component) SyncCommitteeDuties(ctx context.Context, opts *eth2api.SyncCo
 }
 
 func (c Component) Validators(ctx context.Context, opts *eth2api.ValidatorsOpts) (*eth2api.Response[map[eth2p0.ValidatorIndex]*eth2v1.Validator], error) {
-	if len(opts.PubKeys) != 0 {
-		var pubkeys []eth2p0.BLSPubKey
-		for _, pubshare := range opts.PubKeys {
-			pubkey, err := c.getPubKeyFunc(pubshare)
-			if err != nil {
-				return nil, err
-			}
+	if len(opts.PubKeys) == 0 && len(opts.Indices) == 0 {
+		// fetch all validators
+		eth2Resp, err := c.eth2Cl.Validators(ctx, opts)
+		if err != nil {
+			return nil, err
+		}
 
-			pubkeys = append(pubkeys, pubkey)
+		convertedVals, err := c.convertValidators(eth2Resp.Data, len(opts.Indices) == 0)
+		if err != nil {
+			return nil, err
+		}
+
+		return wrapResponse(convertedVals), nil
+	}
+
+	var pubkeys []eth2p0.BLSPubKey
+	for _, pubshare := range opts.PubKeys {
+		pubkey, err := c.getPubKeyFunc(pubshare)
+		if err != nil {
+			return nil, err
 		}
 
-		opts.PubKeys = pubkeys
+		pubkeys = append(pubkeys, pubkey)
 	}
 
-	eth2Resp, err := c.eth2Cl.Validators(ctx, opts)
+	opts.PubKeys = pubkeys
+
+	cachedValidators, err := c.eth2Cl.CompleteValidators(ctx)
 	if err != nil {
 		return nil, err
 	}
-	vals := eth2Resp.Data
 
-	convertedVals, err := c.convertValidators(vals, len(opts.Indices) == 0)
+	cvMap := make(map[eth2p0.BLSPubKey]struct{})
```

Comment:
`cachedValidators` is already a map. Why not traverse `opts.Indices` instead of `opts.PubKeys` to avoid extra maps and loop?

Reply:
`opts.Indices` is an optional field, as is `opts.PubKeys`: not providing them returns the list of all the validators on the network.

Thinking about it some more, we should also handle the caching layer for indices... though, no VC ever queries with it.

Reply:
Understood. Then why valcache uses indices for mapping? Maybe it needs to use pubkeys?

Reply:
> Understood. Then why valcache uses indices for mapping? Maybe it needs to use pubkeys?

I'm caching both the active validator list and the raw Validators response. We cache by index because that's the most common way of looking for validators, although the semantics of the beacon REST API might suggest otherwise.